### PR TITLE
Fix MenuStrip item focus announcement when opening/closing dropdown menu

### LINKED ISSUE #8432: UIA focus is not returned back to menu item after closing dropdown
### .NET version

.NET 8.0.0-alpha.1.22622.1

### Did it work in .NET Framework?

No

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

No

### Issue description

Related to #7770.

After closing MenuStrip's submenu, UIA focus doesn't switch back to the top-level menu item, which may cause issues for screen readers. 

#### Current behavior

In Narrator:
A gif demonstrating the issue when using Narrator: after closing the dropdown via Escape, focus highlight rectangle stays on now invisible menu item

> [Press Alt]
> menuStrip1, Project, menu item, collapsed, 
> [Press Down arrow]
> Project, menu, 
> New, menu item, Ctrl+ N, 
> [Press Escape]

There is no announcement that focus is now on Project menu item after pressing Escape.

In JAWS:
A gif demonstrating the issue when using JAWS: after closing the dropdown via Escape, menu is reopened and announced again as context menu

> [Press Alt]
> Menu bar 
> Project sub menu 
> [Press Down arrow]
> menu 
> New Ctrl+N 
> [Press Escape]
> Context Menu 
> Project 
> menu 
> New Ctrl+N 
> [Press Escape]
> Context Menu 
> Project 
> menu 
> New Ctrl+N

After closing the dropdown, submenu is reopened and announced again as a context menu.

NVDA is similar to Narrator but focus moves to the whole window instead of staying on closed dropdown item.

#### Expected behavior

After dropdown is dismissed, top-level menu item is announced as currently focused item, and the focus highlight moves to it. 

### Steps to reproduce

1. Add a MenuStrip with a menu item with submenu (e.g. via Insert Standard Items button), or open Menus form on WinformsControlsTest app.
2. Start Narrator
3. Expand and close the menu item

### PR DESCRIPTION
**Pull request by v-elnovikova**
Fixes #8432

## Proposed changes

- Fire AutomationFocusChanged UIA-event after dropdown is closed
- When dropdown menu is opened, it's first item is automatically selected and AutomationFocusChanged UIA-event is fired. At this moment, check MenuStrip's AccessibilityObject, not DropDown's, because it might not be created yet

## Customer Impact

- Screen readers announce menu item selection on drop down first opening
- Screen readers announce focus returning back to menu item after dropdown closing

## Regression? 

- No

## Risk

- Minimal

### Before
https://user-images.githubusercontent.com/113603457/211337891-2b890e41-6d62-49cd-ba1a-e8251214c942.mp4

> .NET 8.0.0-alpha.1.23057.5 window, MenuStrip and CheckedListBox window, 
> Exit .NET 8.0.0-alpha.1.23057.5 window, Enter MenuStrip and CheckedListBox window, menuStrip1 menu bar, Project, menu item, collapsed, 
> Project, menu, 
> Check/UnCheck, menu item, checked, Ctrl+ C, 

Missing announcements:
- "New" menu item after opening "Project" menu
- "Project" menu item after closing "Project" menu

### After

https://user-images.githubusercontent.com/113603457/211337951-47c8384e-cfb6-4218-abba-622805667e0b.mp4

> .NET 8.0.0-alpha.1.23057.5 window, MenuStrip and CheckedListBox window, 
> Exit .NET 8.0.0-alpha.1.23057.5 window, Enter MenuStrip and CheckedListBox window, menuStrip1 menu bar, Project, menu item, collapsed, 
> Project, menu, 
> New, menu item, Ctrl+ N, 
> Project, menu, 
> Check/UnCheck, menu item, checked, Ctrl+ C, 
> Project, menu item, collapsed, 

## Test methodology 

- Manual

## Accessibility testing  

Tested in Narrator, NVDA, JAWS
 

## Test environment(s) 

- .NET 8.0.0-alpha.1.23057.5
- Windows 11
- NVDA 2022.3.2
- JAWS 2022 2110.60 ILM

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8473)

### CODE REVIEW

**Comment by v-elnovikova** [comment]
> LGTM! Does this change also improve announcements when navigating submenus like this?

Yes, behavior is the same. Here is an example:

https://user-images.githubusercontent.com/113603457/211378788-56a6d03a-051f-47ce-9819-67bbd797e78a.mp4

**Review by dmitrii-drobotov — DISMISSED** [review]

**Comment by Tanya-Solyanik** [comment]
Looks good to me, thank you for adding videos. Please add an explanation as requested by Devendar

**Review by NikitaSemenovAkvelon — APPROVED** [review]
LGTM

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs [hunk 1]
```diff
@@ -3264,7 +3264,7 @@ public void Select()
 
                 KeyboardToolTipStateMachine.Instance.NotifyAboutGotFocus(this);
 
-                if (IsParentAccessibilityObjectCreated && AccessibilityObject is ToolStripItemAccessibleObject accessibleObject)
+                if ((RootToolStrip?.IsAccessibilityObjectCreated ?? false) && AccessibilityObject is ToolStripItemAccessibleObject accessibleObject)
```
[the remark this answers is not in the document]
Comment (v-elnovikova) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs]: There is no reason to do so. I've replaced RootToolStrip check with parent check. If parent item or ToolStrip has AO instance created, then assume that child item AO should be created too.

**Review by dmitrii-drobotov — APPROVED** [review]
Looks good!

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs [hunk 2]
```diff
@@ -3199,6 +3204,11 @@ internal void RescaleConstantsInternal(int newDpi)
 
         public void Select()
         {
+            Select(false);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs:3207]: please add the parameter name before the literal.

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs [hunk 3]
```diff
@@ -3199,6 +3204,11 @@ internal void RescaleConstantsInternal(int newDpi)
 
         public void Select()
         {
+            Select(forceRaiseA11yFocusChanged: false);
+        }
+
+        internal void Select(bool forceRaiseA11yFocusChanged)
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs:3210]: There are no other usages of `a11y` abbreviation in the codebase, so I would spell it out:
```suggestion
        internal void Select(bool forceRaiseAccessibilityFocusChanged)
```

**Review by dmitrii-drobotov — COMMENTED** [review]
Looks good, only a small nit

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDown.cs [hunk 4]
```diff
@@ -1418,7 +1418,7 @@ internal void SelectPreviousToolStrip()
 
             if (itemOnPreviousMenuToSelect is not null)
             {
-                itemOnPreviousMenuToSelect.Select();
+                itemOnPreviousMenuToSelect.Select(forceRaiseA11yFocusChanged: true);
```
Comment (dkazennov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDown.cs:1421]: `forceRaiseA11yFocusChanged` doesn't make sense to me and I haven't found any other use of this abbreviation (or similar one) either. Just `forceRaiseAOFocusChanged` or `forceRaiseAccessibilityObjectFocusChanged` may be better.

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs (hunk 3 above)
Comment (v-elnovikova) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs]: Renamed to `forceRaiseAccessibilityFocusChanged` as less confusing

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDown.cs (hunk 4 above)
Comment (v-elnovikova) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDown.cs]: Renamed to `forceRaiseAccessibilityFocusChanged`

**Review by dmitrii-drobotov — APPROVED** [review]

**Review by Tanya-Solyanik — APPROVED** [review]

**Pull request merged**